### ISSUE TITLE
Add notes on contributing, especially w.r.t the parser verification check

### ISSUE THREAD

**Issue by fwcd** [fwcd/tree-sitter-kotlin#92]
New PRs regularly fail the CI check that verifies that the generated parser matches the committed `parser.c`, presumably because their development machine didn't generate a byte-for-byte-exact copy of the parser as the GitHub Actions runner.

Perhaps we could add some notes to clarify the motivation behind this check and how to successfully generate a conforming `parser.c`. Something along the lines of:

- The check exists since it is infeasible to review a ~20 MB C source file. Suggestions for something less brittle than a byte-for-byte check are of course welcome, however.
- Unfortunately, the output of `tree-sitter generate` only seems to be deterministic when run on the same platform. Since the GitHub runner use x86_64 Linux, contributors have to generate the parser from that platform too (even aarch64 Linux/macOS will produce a different output).
- One option besides running Linux natively or e.g. through a Docker container would be to use the `regenerate-parser.yml` workflow in their own fork. This would potentially require customizing the branch it runs on.

**Comment by ketkarameya** [comment]
Do we really need `parser.c` in the repo. The tree-sitter-swift or tree-sitter-java don't have it in the repo either. 

https://github.com/alex-pinkus/tree-sitter-swift#frequently-asked-questions
https://github.com/tree-sitter/tree-sitter-java

**Comment by fwcd** [comment]
`tree-sitter-java` does in fact check it in ([see here](https://github.com/tree-sitter/tree-sitter-java/blob/master/src/parser.c)), but I do like the approach of `tree-sitter-swift`, that is, omitting the parser and publishing it as a CI artifact. That would probably also help with repo growth, which may become problematic due to the frequent changes in `parser.c`.

The only reason I am a bit hesitant is that it seems to deviate from the dominant convention of including the parser, even if it's comparatively large. See e.g.

- [`tree-sitter-c-sharp` (41 MB)](https://github.com/tree-sitter/tree-sitter-c-sharp/blob/master/src/parser.c)
- [`tree-sitter-ocaml` (32 MB)](https://github.com/tree-sitter/tree-sitter-ocaml/blob/694c57718fd85d514f8b81176038e7a4cfabcaaf/ocaml/src/parser.c)
- [`tree-sitter-haskell` (23 MB)](https://github.com/tree-sitter/tree-sitter-haskell/blob/ba0bfb0e5d8e9e31c160d287878c6f26add3ec08/src/parser.c)
- [`tree-sitter-scala` (23 MB)](https://github.com/tree-sitter/tree-sitter-scala/blob/8062487fb3b7f3ce1bb7ce1fd1c84bed60c75203/src/parser.c)
- [`tree-sitter-cpp` (16 MB)](https://github.com/tree-sitter/tree-sitter-cpp/blob/cbb9974efde90cbdc5d1027c97882df737aa3601/src/parser.c)
- [`tree-sitter-ruby` (14 MB)](https://github.com/tree-sitter/tree-sitter-ruby/blob/f257f3f57833d584050336921773738a3fd8ca22/src/parser.c)
- [`tree-sitter-agda` (13 MB)](https://github.com/tree-sitter/tree-sitter-agda/blob/ca69cdf485e9ce2b2ef0991a720aa88d87d30231/src/parser.c)

**Comment by fwcd** [comment]
Might be worth keeping an eye on

- https://github.com/tree-sitter/tree-sitter/issues/930
- https://github.com/tree-sitter/tree-sitter/issues/730

which propose to remove the generated files from the repos.

**Comment by clason** [comment]
...but also include it as a release artefact for semantic versioned(!) releases. (This parser takes a comparatively long time to generate, so just dropping it would not be preferred albeit workable. A separate "release branch", though, is highly discouraged -- some parsers chose to do that in the early days, but it's more headache for everyone than it's worth.)

In either case, please keep committing the generated json, as these are enough to generate the parser without node. (Not that your parser requires `npm`, as far as I can tell, but it's the principle.)

**Comment by fwcd** [comment]
Yes, to be clear, I don't propose to remove anything for now, just to monitor what upstream does. When the official recommendation changes, we can revisit this.

**Comment by clason** [comment]
Sure; I was just explaining current upstream policy ;)
